### DISCUSSION TITLE
reuse EVP_CIPHER_CTX with IV generated internally cause error because of EVP_EncryptInit_ex2(ctx, 0, 0, 0, 0)-->gcm_init(7) does NOT clear PROV_GCM_CTX::iv_state

### DISCUSSION THREAD

**Discussion by PM-Zefman-R** [openssl/openssl#22463] [Q&A]
Hi there.
I'm new to openssl. I wrote a simple test application based on a demo.
I use AES in GCM mode with IV generated internally.
And I'm trying to reuse EVP_CIPHER_CTX - I call EVP_CIPHER_CTX_new() only twice for encoder and decoder operations on the same key.
Each encode operation starts with EVP_EncryptInit_ex2(ctx_enc_, 0, 0, 0, 0);
The first encode|decode operations work well.
But second encode fails on EVP_EncryptUpdate(ctx_enc_, 0, ...) (set AAD) with error:  "AC360000:error:1C800066:Provider routines:ossl_gcm_stream_update:cipher operation failed:providers\implementations\ciphers\ciphercommon_gcm.c:320:"

I think it is all because of EVP_EncryptInit_ex2(ctx_enc_, 0, 0, 0, 0) --> gcm_init(7); does NOT clear PROV_GCM_CTX::iv_state to IV_STATE_UNINITIALISED. It stays IV_STATE_FINISHED (just a thought, not checked as I use openssl as a library).

So the questions are:
- Is it an error that EVP_EncryptInit_ex2(ctx_enc_, 0, 0, 0, 0) --> gcm_init(7) does NOT clear PROV_GCM_CTX::iv_state?
- How can I clear PROV_GCM_CTX::iv_state manually?

**Comment by paulidale** [comment]
I'm not convinced this is a bug in `EVP_EncryptInit_ex2()` even though it's kind of odd.  Reusing the IV in GCM mode is a critical error that completely compromises confidentiality.  It's very very bad.

You should set a **new** IV as part of the call.  That should fix the state.

**Comment by PM-Zefman-R** [comment]
As it is said in manual on page [https://www.openssl.org/docs/man3.1/man3/EVP_EncryptInit_ex2.html] : "For EVP_CIPH_GCM_MODE the IV will be generated internally if it is not specified." So, I just want IV to be generated internally, but it only works for the first cipher operation (EVP_EncryptInit_ex2(); EVP_EncryptUpdate(); EVP_EncryptUpdate(); EVP_EncryptFinal_ex()).
Yes, if I manually specify IV in EVP_EncryptInit_ex2() --> gcm_init() if (iv != NULL) { ... ctx->iv_state = IV_STATE_BUFFERED;} then everything works fine. So, for now I generate IV by myself as a work around. But it seems for me that EVP_EncryptInit_ex2() in case of NULL iv MUST clear iv_state to IV_STATE_UNINITIALISED.

**Comment by pedrolamarao** [comment]
I have just been hit by this bug. Our automated testing was designed to reuse a small set of pre-encrypted objects with pregenerated keys and parameters. We recently added support for AES/GCM and testing fails in every situation our product reuses contexts and our testing reuses keys and ivs. We are unsure if the argument "this is a feature because GCM is unsafe when key/iv pairs are reused" is valid.

**Reply by paulidale** [reply]
I do not think that _unsafe_ is the correct word to use here.  _completely, totally and utterly broken_ or _100% compromised_ are more accurate.  GCM IV reused is a critical mistake.  There is no way the project would accept a change that _fixed_ this.

I.e. fix your tests.

**Reply by pedrolamarao** [reply]
We did, we added the reinitialization of the context with the same fixed IV. The library does, of course, accept it. So, irrespective of what is the appropriate use of GCM in production, this "feature" is no "feature" at all, it is just an implementation quirk. We have quickly moved on.
